### PR TITLE
Make object fields static, and move post-super init to <clinit>

### PR DESCRIPTION
**Pull request by retronym**
This lets us make the MODULE$ field unconditionally final,
which improves runtime performance marginally by letting
JIT elide null checks.

Fields of objects themselves and now also be JIT-inlinable
constants.

Fixes scala/scala-dev#537

### CODE REVIEW

**Comment by retronym** [comment]
TODO:
 - [ ] Community build
 - [ ] Let's try to foresee problems with reflection, framework code, tooling. E.g. would scalamock still work?
 - [ ] Should we offer the legacy encoding with a compiler option?

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala:122]: @sjrd Thoughts?

Comment (sjrd) [reply src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala:122]: I haven't made my mind about it yet. I'll think about it.

Can my input wait until next week? I'm currently at the Scala Symposium.

Comment (retronym) [reply src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala:122]: Sure thing.

**Comment by retronym** [comment]
There is one interesting failure in `test/files/jvm/scala-concurrent-tck.scala` that shows a subtle difference in locking during object initialization.

Minimized:

```scala
trait CustomExecutionContext {
  val inEC = ""

  def testOnSuccessCustomEC(): Unit = {
    val t = new Thread() {
      override def run(): Unit = {
        println("getting it")
        inEC.length
        // this.outer$.inEC() now blocks until the `Test$.<clinit>` has concluded, because `Test$.inEC()`
        // now differs:
        // public inEC()Ljava/lang/String;
        // -    ALOAD 0
        // -    GETFIELD Test$.inEC : Ljava/lang/String;
        // +    GETSTATIC Test$.inEC : Ljava/lang/String;
        //
        // And the GETSTATIC blocks.
        println("got it")
      }
    }
    t.start()
    t.join()
    println("done")
  }

  testOnSuccessCustomEC()
}

object Test extends CustomExecutionContext {
  def main(args: Array[String]): Unit = {
  }
}

```

This is reminiscent of the deadlocks that became possible with the 2.12 lambda encoding. Prior to that, lambdas were inner classes, which had access to the module instance field via the outer pointer.

Now, inner classes, even though they have the module instance, are exposed to the static init lock when the call a field accessor method.

**Comment by retronym** [comment]
Another thing that's changed in the serialized form of top level modules:

```
$ qscala
Welcome to Scala 2.13.0-20180928-004946-4aa133f (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> import java.util.Base64; def parseBase64Binary(s: String): Array[Byte] = Base64.getDecoder.decode(s); def printBase64Binary(data: Array[Byte]): String = Base64.getEncoder.encode(data).map(_.toChar).mkString; def serialize(o: AnyRef): String = {val bos = new java.io.ByteArrayOutputStream();val out = new java.io.ObjectOutputStream(bos);out.writeObject(o);out.flush();printBase64Binary(bos.toByteArray())}; serialize(List)
warning: -nc is deprecated: scripts use cold compilation by default; use -Yscriptrunner for custom behavior
import java.util.Base64
parseBase64Binary: (s: String)Array[Byte]
printBase64Binary: (data: Array[Byte])String
serialize: (o: AnyRef)String
res0: String = rO0ABXNyACBzY2FsYS5jb2xsZWN0aW9uLmltbXV0YWJsZS5MaXN0JAAAAAAAAAADAwAAeHB4

scala> :quit
~/code/scala on topic/static-object-fields*
$ scala-ref 2.13.x
Welcome to Scala 2.13.0-20180927-195543-c277d7d (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> import java.util.Base64; def parseBase64Binary(s: String): Array[Byte] = Base64.getDecoder.decode(s); def printBase64Binary(data: Array[Byte]): String = Base64.getEncoder.encode(data).map(_.toChar).mkString; def serialize(o: AnyRef): String = {val bos = new java.io.ByteArrayOutputStream();val out = new java.io.ObjectOutputStream(bos);out.writeObject(o);out.flush();printBase64Binary(bos.toByteArray())}; serialize(List)
import java.util.Base64
parseBase64Binary: (s: String)Array[Byte]
printBase64Binary: (data: Array[Byte])String
serialize: (o: AnyRef)String
res0: String = rO0ABXNyACBzY2FsYS5jb2xsZWN0aW9uLmltbXV0YWJsZS5MaXN0JAAAAAAAAAADAwABTAARcGFydGlhbE5vdEFwcGxpZWR0ABFMc2NhbGEvRnVuY3Rpb24xO3hweA==
```

This confuses me. The following details of `List$` are the same before and after this PR.

```
public final class scala/collection/immutable/List$ implements scala/collection/StrictOptimizedSeqFactory  {

  // access flags 0x1A
  private final static J serialVersionUID = 3

  // access flags 0x2
  private writeObject(Ljava/io/ObjectOutputStream;)V
    // parameter final  out
    RETURN
    MAXSTACK = 0
    MAXLOCALS = 2


  // access flags 0x2
  private readObject(Ljava/io/ObjectInputStream;)V
    // parameter final  in
    RETURN
    MAXSTACK = 0
    MAXLOCALS = 2

  // access flags 0x2
  private readResolve()Ljava/lang/Object;
    GETSTATIC scala/collection/immutable/List$.MODULE$ : Lscala/collection/immutable/List$;
    ARETURN
    MAXSTACK = 1
    MAXLOCALS = 1
```

Aren't these enough to keep serialization stable?

**Comment by retronym** [comment]
Ah, looking at the implementation of serialization, it records the descriptors of instance fields, even when `serialVersionUID` is explicitly specified. However, when deserializing, if the `svUID` in the data stream matches that of the current version of the class, the field descriptors don't have to align.

```
$ qscala
Welcome to Scala 2.13.0-20180928-004946-4aa133f (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> import java.util.Base64; def parseBase64Binary(s: String): Array[Byte] = Base64.getDecoder.decode(s); def printBase64Binary(data: Array[Byte]): String = Base64.getEncoder.encode(data).map(_.toChar).mkString; def serialize(o: AnyRef): String = {val bos = new java.io.ByteArrayOutputStream();val out = new java.io.ObjectOutputStream(bos);out.writeObject(o);out.flush();printBase64Binary(bos.toByteArray())}; def deserialize(string: String): AnyRef = {val bis = new java.io.ByteArrayInputStream(parseBase64Binary(string)); val in = new java.io.ObjectInputStream(bis); in.readObject()}; serialize(List);
warning: -nc is deprecated: scripts use cold compilation by default; use -Yscriptrunner for custom behavior
import java.util.Base64
parseBase64Binary: (s: String)Array[Byte]
printBase64Binary: (data: Array[Byte])String
serialize: (o: AnyRef)String
deserialize: (string: String)AnyRef
res0: String = rO0ABXNyACBzY2FsYS5jb2xsZWN0aW9uLmltbXV0YWJsZS5MaXN0JAAAAAAAAAADAwAAeHB4

scala> deserialize("rO0ABXNyACBzY2FsYS5jb2xsZWN0aW9uLmltbXV0YWJsZS5MaXN0JAAAAAAAAAADAwABTAARcGFydGlhbE5vdEFwcGxpZWR0ABFMc2NhbGEvRnVuY3Rpb24xO3hweA")
warning: -nc is deprecated: scripts use cold compilation by default; use -Yscriptrunner for custom behavior
res2: AnyRef = scala.collection.immutable.List$@a9f023e
```

```
$ scala-ref 2.13.x
Welcome to Scala 2.13.0-20180927-195543-c277d7d (OpenJDK 64-Bit Server VM, Java 1.8.0-adoptopenjdk).
Type in expressions for evaluation. Or try :help.

scala> import java.util.Base64; def parseBase64Binary(s: String): Array[Byte] = Base64.getDecoder.decode(s); def printBase64Binary(data: Array[Byte]): String = Base64.getEncoder.encode(data).map(_.toChar).mkString; def serialize(o: AnyRef): String = {val bos = new java.io.ByteArrayOutputStream();val out = new java.io.ObjectOutputStream(bos);out.writeObject(o);out.flush();printBase64Binary(bos.toByteArray())}; def deserialize(string: String): AnyRef = {val bis = new java.io.ByteArrayInputStream(parseBase64Binary(string)); val in = new java.io.ObjectInputStream(bis); in.readObject()}; serialize(List);
import java.util.Base64
parseBase64Binary: (s: String)Array[Byte]
printBase64Binary: (data: Array[Byte])String
serialize: (o: AnyRef)String
deserialize: (string: String)AnyRef
res0: String = rO0ABXNyACBzY2FsYS5jb2xsZWN0aW9uLmltbXV0YWJsZS5MaXN0JAAAAAAAAAADAwABTAARcGFydGlhbE5vdEFwcGxpZWR0ABFMc2NhbGEvRnVuY3Rpb24xO3hweA==

scala> deserialize("rO0ABXNyACBzY2FsYS5jb2xsZWN0aW9uLmltbXV0YWJsZS5MaXN0JAAAAAAAAAADAwAAeHB4")
res2: AnyRef = scala.collection.immutable.List$@595f9916
```

So functionally, we have forwards and backwards serialization compatibility, but the irrelevent deatils in the serialization format make it hard for our tests to show that.

An alternative would be to serialize all top level modules `writeReplace`/`readResolve` to a class that reflectively accessed the `MODULE$` field on deserialization. For performance, it could cache that lookup in a `ClassValue`.

```java
package scala.runtime;

public final class SerializedObject {
    private final Class<?> moduleClass;
    private static final ClassValue<Object> instances = new ClassValue<Object>() {
        @Override
        protected Object computeValue(Class<?> type) {
            try {
                return type.getField("MODULE$").get(null);
            } catch (IllegalAccessException | NoSuchFieldException e) {
                throw new RuntimeException(e);
            }
        }
    };

    public SerializedObject(Class<?> moduleClass) {
        this.moduleClass = moduleClass;
    }

    private Object readResolve() {
        return instances.get(moduleClass);
    }
}

```

**Comment by retronym** [comment]
Here's a demo that shows that object val members can now be JIT inlined:

```scala
import java.lang.invoke.{MethodHandle, MethodHandles, MethodType}

class C

object Test {
  val handle = MethodHandles.lookup().findVirtual(Test.getClass, "callMe", MethodType.methodType(classOf[Object]))

  def newObject2 = handle.invoke(this)

  @volatile var z = 0

  def main(args: Array[String]): Unit = {
    while(true) {
      z += newObject2.hashCode()
    }
  }

  def callMe = new Object
}
```

```
$ qscalac sandbox/test.scala && jardiff 'Test$.class' && (jabba use openjdk@1.11.0 && qscala -J-XX:+UnlockDiagnosticVMOptions -J-XX:+PrintInlining Test)

+++ Test$.class.asm
// class version 52.0 (52)
// access flags 0x31
public final class Test$ {

  // access flags 0x19
  public final static INNERCLASS java/lang/invoke/MethodHandles$Lookup java/lang/invoke/MethodHandles Lookup

  // access flags 0x19
  public final static LTest$; MODULE$

  // access flags 0x1A
  private final static Ljava/lang/invoke/MethodHandle; handle

  // access flags 0x4A
  private static volatile I z

  // access flags 0x9
  public static <clinit>()V
    NEW Test$
    DUP
    INVOKESPECIAL Test$.<init> ()V
    PUTSTATIC Test$.MODULE$ : LTest$;
    INVOKESTATIC java/lang/invoke/MethodHandles.lookup ()Ljava/lang/invoke/MethodHandles$Lookup;
    GETSTATIC Test$.MODULE$ : LTest$;
    INVOKEVIRTUAL Test$.getClass ()Ljava/lang/Class;
    LDC "callMe"
    LDC Ljava/lang/Object;.class
    INVOKESTATIC java/lang/invoke/MethodType.methodType (Ljava/lang/Class;)Ljava/lang/invoke/MethodType;
    INVOKEVIRTUAL java/lang/invoke/MethodHandles$Lookup.findVirtual (Ljava/lang/Class;Ljava/lang/String;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/MethodHandle;
    PUTSTATIC Test$.handle : Ljava/lang/invoke/MethodHandle;
    ICONST_0
    PUTSTATIC Test$.z : I
    RETURN
    MAXSTACK = 4
    MAXLOCALS = 0

  // access flags 0x2
  private <init>()V
    ALOAD 0
    INVOKESPECIAL java/lang/Object.<init> ()V
    RETURN
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public callMe()Ljava/lang/Object;
    NEW java/lang/Object
    DUP
    INVOKESPECIAL java/lang/Object.<init> ()V
    ARETURN
    MAXSTACK = 2
    MAXLOCALS = 1

  // access flags 0x1
  public handle()Ljava/lang/invoke/MethodHandle;
    GETSTATIC Test$.handle : Ljava/lang/invoke/MethodHandle;
    ARETURN
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public main([Ljava/lang/String;)V
    // parameter final  args
   L0
    ALOAD 0
    ALOAD 0
    INVOKEVIRTUAL Test$.z ()I
    ALOAD 0
    INVOKEVIRTUAL Test$.newObject2 ()Ljava/lang/Object;
    INVOKEVIRTUAL java/lang/Object.hashCode ()I
    IADD
    INVOKEVIRTUAL Test$.z_$eq (I)V
    GOTO L0
    MAXSTACK = 3
    MAXLOCALS = 2

  // access flags 0x1
  public newObject2()Ljava/lang/Object;
    ALOAD 0
    INVOKEVIRTUAL Test$.handle ()Ljava/lang/invoke/MethodHandle;
    ALOAD 0
    INVOKEVIRTUAL java/lang/invoke/MethodHandle.invoke (LTest$;)Ljava/lang/Object;
    ARETURN
    MAXSTACK = 2
    MAXLOCALS = 1

  // access flags 0x1
  public z()I
    GETSTATIC Test$.z : I
    IRETURN
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public z_$eq(I)V
    // parameter final  x$1
    ILOAD 1
    PUTSTATIC Test$.z : I
    RETURN
    MAXSTACK = 1
    MAXLOCALS = 2
}
...
                                @ 1   Test$::handle (4 bytes)   inline (hot)
                                @ 5   java.lang.invoke.LambdaForm$MH/0x00000008001a1840::invoke_MT (22 bytes)   force inline by annotation
                                  @ 8   java.lang.invoke.Invokers::checkGenericType (6 bytes)   force inline by annotation
                                    @ 2   java.lang.invoke.MethodHandle::asType (28 bytes)   inline (hot)
                                  @ 13   java.lang.invoke.Invokers::checkCustomized (23 bytes)   force inline by annotation
                                    @ 1   java.lang.invoke.MethodHandleImpl::isCompileConstant (2 bytes)   (intrinsic)
                                  @ 18   java.lang.invoke.LambdaForm$DMH/0x00000008001a1c40::invokeSpecial (20 bytes)   force inline by annotation
                                    @ 7   java.lang.invoke.DirectMethodHandle::internalMemberName (8 bytes)   force inline by annotation
                                    @ 16   Test$::callMe (8 bytes)   inline (hot)
                                      @ 4   java.lang.Object::<init> (1 bytes)   inline (hot)
```

Before:

```
                              @ 6   Test$::newObject2 (9 bytes)
                                @ 1   Test$::handle (5 bytes)
                                @ 5   java.lang.invoke.LambdaForm$MH/0x00000008001d6040::invoke_MT (22 bytes)   force inline by annotation
                                  @ 8   java.lang.invoke.Invokers::checkGenericType (6 bytes)   force inline by annotation
                                    @ 2   java.lang.invoke.MethodHandle::asType (28 bytes)   callee is too large
                                  @ 13   java.lang.invoke.Invokers::checkCustomized (23 bytes)   force inline by annotation
                                    @ 1   java.lang.invoke.MethodHandleImpl::isCompileConstant (2 bytes)
                                    @ 19   java.lang.invoke.Invokers::maybeCustomize (28 bytes)   don't inline by annotation
                                  @ 18   java.lang.invoke.MethodHandle::invokeBasic(L)L (0 bytes)   receiver not constant
                              @ 9   java.lang.Object::hashCode (0 bytes)   no static binding
```

**Comment by retronym** [comment]
> An alternative would be to serialize all top level modules ... [with a serialization proxy]

I see @szeiger has proposed this already in https://github.com/scala/scala/pull/6877

**Comment by retronym** [comment]
Benchmarks points to [appear here](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&var-scalaSha=6c8459bd60%7Cf904ba94d5%7Cc277d7d7ca%7C0f926ac098%7C54771e292f%7Ccba48870bd%7Ca9934ad6ae%7Ca0837c220e%7Ca2b0d26c91%7C80080695b3%7C2e82444401%7C6266879651%7C2e5399c382%7Cbcc5c7c13d%7C35f71fa3ef%7C3826f6e5ac%7C1c2d4ad3eb%7Cb9eab9cfb4%7C75c8fe1720%7Cb75bfc3b78%7Cafaea98132%7C1112dd1661%7C3aa6273ac9%7C0bd93be8c4%7C412af9c2aa%7C2e385f9ea3%7C9bfd7d82c2%7C84de40d84e%7C465aea6695%7C528adb03ff%7C4c9136a61c%7Ce265e4278a&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40&from=1536469131959&to=1538117096930).

The first will show the improvement from using this PR as STARR, rather than the preceding commit. The second point will show the subsequent change from using this PR as STARR and this PR as the compiler-under-performance test, to see if this extra work in the backend makes a difference.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala:133]: It's maybe a little confusing to create the field symbol but not enter it in the decls and rely on `addModuleInstanceField`. A comment would be helpful.

(Class fields are generated based on `BCodeHelpers.fieldSymbols`, so on the `decls`. Methods on the other hand are generated based on `DefDef`s.)

File: test/files/jvm/scala-concurrent-tck.scala
Comment (viktorklang) [inline test/files/jvm/scala-concurrent-tck.scala:40]: @retronym Out of curiosity, why did these need to change?

Comment (retronym) [reply test/files/jvm/scala-concurrent-tck.scala:40]: https://github.com/scala/scala/pull/7270#issuecomment-425275856

As written, the tests ran during the `Test$.<clinit>`, during which time they started threads that called back into accessors of mixed-in fields (like `inEc`) in `Test$`. That didn't deadlock because those fields were non-static and the reference wasn't made via the static `Test$.MODULE$`, but rather through the `outer$` reference of an inner class in the mixed-in trait.

After this PR, pattern leads to a deadlock: the `<clinit>` thread is waiting for the other thread to stop, and the other thread is blocked on the `GETSTATIC` instruction.

Comment (viktorklang) [reply test/files/jvm/scala-concurrent-tck.scala:40]: @retronym That makes me feel skeptical about this encoding. It would be interesting to see the community build running on this PR.

Comment (retronym) [reply test/files/jvm/scala-concurrent-tck.scala:40]: Yeah, it's unfortunate. But code that publishes `this` for use on other threads before initialization is complete is asking for trouble, and tends to run afoul of subtle differences in the way things are compiled. We had the same problem with the lambda encoding on 2.12.

Comment (viktorklang) [reply test/files/jvm/scala-concurrent-tck.scala:40]: @retronym Let's see what community-build says :)

File: src/reflect/scala/reflect/internal/TreeInfo.scala
Comment (viktorklang) [inline src/reflect/scala/reflect/internal/TreeInfo.scala:1097]: This seems like a lot of iteration and a lot of allocation. Would this not be possible to solve with a custom Ordering + a sort?

Comment (retronym) [reply src/reflect/scala/reflect/internal/TreeInfo.scala:1097]: Yes it could be done with a bit less garbage, but this isn't a hotspot based on profiles I've collected.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala]: Maybe crate a `nme.CLASS_CONSTRUCTOR`, or use `GenBCode.CLASS_CONSTRUCTOR_NAME`

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala:125]: It's a little bit funny that this is enough to make code gen emit `getstatic/putstatic` and avoid loading the module instance. The trees accessing the fields still have the shape `Select(This("T"), "y")`. It's fine with me; we don't really have a "correct" way to represent a static selection in trees, I think we discussed this before.

File: src/reflect/scala/reflect/internal/TreeInfo.scala
Comment (viktorklang) [reply src/reflect/scala/reflect/internal/TreeInfo.scala:1097]: It's your call :) I always err on the side of efficiency :)

**Review by lrytz — COMMENTED** [review]
If the object inherits `val` fields from traits, the `<clinit>` now gets a `runtime.Statics.releaseFence()` call. Given the class init lock, this is redundant, no?

The changes LGTM. Let's see what others say (and the community build).

**Comment by smarter** [comment]
I asked Dmitry about this encoding since he spent some time on this problem a while ago, he said:
> The scheme makes sense.
I'd tune it a bit to change behaviour of inner classes of modules, so that 
they fall back to outer references. I think this will allow to optimise the 
behaviour by JIT, while keeping current semantics.

File: build.sbt
Comment (som-snytt) [inline build.sbt]: is this a duplicated line? My eyes just blurred over.

**Comment by retronym** [comment]
Rebased on #7297, which locks down the serialized form of top level modules, making it insensitive to the changes to field modifiers.

File: build.sbt
Comment (retronym) [reply build.sbt]: Thanks, I've removed the offending ⌘V.

**Comment by retronym** [comment]
> making it insensitive to the changes to field modifiers.

Bzzzt, wrong again, me. The java serialiazation for `List$.class` actually changes depending on what serializable (non-static/transient) fields it contains! So this pull request does change the serialized form, by changing `List$.partialNotApplied` to static. Such a change is forwards/backwards serialization compatible (not that we aspire to that until we lock down 2.13.0.)

;tl;dr Going forwards, we will no longer incur checkfile updates to the serialization stability test when we add or remove fields. But we need one last update now!

**Comment by SethTisue** [comment]
community build run https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1491/consoleFull

**Comment by retronym** [comment]
Rebased.

**Comment by retronym** [comment]
There appear to be unrelated failures in the community build run above related to collections changes and scala-java8-compat.

I've re-run as https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1522/ after rebasing this PR.

I've also prepared a [backport to 2.12.x](https://github.com/scala/scala/compare/2.12.x...retronym:backport/static-object-fields?expand=1) to facilitate testing with a cleaner, and more comphrehensive (?) community build.

**Comment by SethTisue** [comment]
> There appear to be unrelated failures in the community build run above related to collections changes and scala-java8-compat

confirm, unrelated

**Comment by mkeskells** [comment]
is there a performance run with this PR?
I have a couple of branches that should benefit from this change that I will hope to benchmark this week

**Comment by mkeskells** [comment]
is there a performance run with this PR?
I have a couple of branches that should benefit from this change that I will hope to benchmark this week

**Comment by mkeskells** [comment]
Is this PR ready to merge. It seems to have gone quiet

Are we waiting for some additional review or tests or something else?

**Comment by lrytz** [comment]
@retronym did you see / address my comment?

> If the object inherits `val` fields from traits, the `<clinit>` now gets a `runtime.Statics.releaseFence()` call. Given the class init lock, this is redundant, no?

**Comment by retronym** [comment]
I saw that but forgot about it. I'll see what I can do about that.

**Comment by lrytz** [comment]
Maybe add a `nsc.backend.jvm.BytecodeTest` for the last commit?

**Comment by retronym** [comment]
Test added, and rebased.

**Comment by lrytz** [comment]
Hmm, test failure after the rebase.. https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/5471/testReport/junit/test.files/pos/inline_access_levels/

**Comment by retronym** [comment]
@lrytz Looks like [this TODO](https://github.com/scala/scala/blob/c5c28ebf304cb2e50022c0f307c2d9ad7d06505c/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala#L343-L349) in the inliner heuristics about identifying potential private field access is biting us now that a `GETSTATIC` can reference a private field.

I tried:

```diff
diff --git a/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala b/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala
index b0461aa2a7..9c52b2d385 100644
--- a/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala
+++ b/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala
@@ -21,7 +21,7 @@ import scala.annotation.{switch, tailrec}
 import scala.collection.JavaConverters._
 import scala.collection.immutable.BitSet
 import scala.collection.mutable
-import scala.reflect.ClassTag
+import scala.reflect.{ClassTag, NameTransformer}
 import scala.reflect.internal.util.Position
 import scala.tools.asm
 import scala.tools.asm.Opcodes._
@@ -379,7 +379,14 @@ abstract class BackendUtils extends PerRunInit {
             callMi = mi
           }
         }
-      } else if (i.getOpcode != GETSTATIC && nonForwarderInstructionTypes(t))
+      } else if (i.getOpcode == GETSTATIC) {
+        val name = i.asInstanceOf[FieldInsnNode].name
+        val owner = i.asInstanceOf[FieldInsnNode].owner
+        if (name != NameTransformer.MODULE_INSTANCE_NAME || owner == method.) {
+          // println(owner + " " + name)
+          numCallsOrNew = 2 // stop here: not forwarder or trivial
+        }
+      } else if (nonForwarderInstructionTypes(t))
         numCallsOrNew = 2 // stop here: not forwarder or trivial
     }
     if (numCallsOrNew > 1 || numBoxConv > paramTypes.length + 1) -1

```

But that triggers failures in `scala.tools.nsc.backend.jvm.opt.InlinerTest.optimizeSpecializedClosures` due to failure to inline a method with a `GETSTATIC` on either `java/lang/Byte$ByteCache cache` and/or `scala/runtime/BoxedUnit UNIT`

**Comment by retronym** [comment]
This variation makes that test pass again, but I'm not happy enough to submit it.

```diff
diff --git a/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala b/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala
index b0461aa2a7..4be28895f7 100644
--- a/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala
+++ b/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala
@@ -21,7 +21,7 @@ import scala.annotation.{switch, tailrec}
 import scala.collection.JavaConverters._
 import scala.collection.immutable.BitSet
 import scala.collection.mutable
-import scala.reflect.ClassTag
+import scala.reflect.{ClassTag, NameTransformer}
 import scala.reflect.internal.util.Position
 import scala.tools.asm
 import scala.tools.asm.Opcodes._
@@ -379,7 +379,15 @@ abstract class BackendUtils extends PerRunInit {
             callMi = mi
           }
         }
-      } else if (i.getOpcode != GETSTATIC && nonForwarderInstructionTypes(t))
+      } else if (i.getOpcode == GETSTATIC) {
+        val name = i.asInstanceOf[FieldInsnNode].name
+        val owner = i.asInstanceOf[FieldInsnNode].owner
+        if (name != NameTransformer.MODULE_INSTANCE_NAME && !(owner == "scala/runtime/BoxedUnit" && name == "UNIT")) {
+          // println(owner + " " + name)
+          // println(AsmUtils.textify(method))
+          numCallsOrNew = 2 // stop here: not forwarder or trivial
+        }
+      } else if (nonForwarderInstructionTypes(t))
         numCallsOrNew = 2 // stop here: not forwarder or trivial
     }
     if (numCallsOrNew > 1 || numBoxConv > paramTypes.length + 1) -1

```

**Comment by retronym** [comment]
@lrytz I've pushed a change that only allows `GETSTATIC`-s of fields from classes that differ from the methods owner.

**Review by lrytz — APPROVED** [review]
👍 Thanks!

**Pull request merged**
